mllib/src/main/scala/org/apache/spark/mllib/optimization/GradientDescent.scala

```diff
@@ -219,4 +265,39 @@ object GradientDescent extends Logging {
     (weights, stochasticLossHistory.toArray)
 
   }
+
+  def runMiniBatchSGD(
+      data: RDD[(Double, Vector)],
+      gradient: Gradient,
+      updater: Updater,
+      stepSize: Double,
+      numIterations: Int,
+      regParam: Double,
+      miniBatchFraction: Double,
+      initialWeights: Vector): (Vector, Array[Double]) =
+    GradientDescent.runMiniBatchSGD(data, gradient, updater, stepSize, numIterations,
+                                    regParam, miniBatchFraction, initialWeights, 0.001)
+
+
+  private def isConverged(previousWeights: Vector, currentWeights: Vector,
+                          initialWeights: Vector, convergenceTol: Double): Boolean = {
+    require(previousWeights != None)
+    require(currentWeights != None)
+    // To compare with convergence tolerance
+    def solutionVecDiff(previousWeight: Vector,
+                        currentWeight: Vector): Double = {
+
+      val lastWeight = currentWeight.toBreeze
+      val lastBeforeWeight = previousWeight.toBreeze
+      sum((lastBeforeWeight - lastWeight)
+        :* (lastBeforeWeight - lastWeight)) / lastWeight.length
+    }
+
+    def squareAvg(weights: Vector): Double =
+      sum(weights.toBreeze :* weights.toBreeze) / weights.toBreeze.length
+
+    solutionVecDiff(previousWeights, currentWeights) <
+      convergenceTol * squareAvg(initialWeights)
```

Comment:
I believe Breeze has norm() functions you can use instead.
This norm should be precomputed and stored.